src/Core/LanguageService.fs

```diff
@@ -732,64 +735,51 @@ Consider:
-                        // else
-                        //     return [], [], fsacPath
-                        }
-
-                    let userDotnetArgs = "FSharp.fsac.dotnetArgs" |> Configuration.get [||]
-
-                    let hasUserRollForward =
-                        userDotnetArgs
-                        |> Array.tryFindIndex (fun a -> a = "--roll-forward")
-                        |> Option.map (fun _ -> true)
-                        |> Option.defaultValue false
-
-                    let hasUserFxVersion =
-                        userDotnetArgs
-                        |> Array.tryFindIndex (fun a -> a = "--fx-version")
-                        |> Option.map (fun _ -> true)
-                        |> Option.defaultValue false
-
-                    let shouldApplyImplicitRollForward = not (hasUserFxVersion || hasUserRollForward)
 
-                    let args =
-                        [ if shouldApplyImplicitRollForward then
-                              yield! rollForwardArgs
-                          yield! userDotnetArgs ]
-
-                    let envVariables =
-                        [ if shouldApplyImplicitRollForward then
-                              yield! necessaryEnvVariables ]
-
-                    return args, envVariables, fsacPath
+                    let! sdkVersionAtRootPath = sdkVersion ()
+
+                    match sdkVersionAtRootPath with
+                    | Error e ->
+                        printfn $"Error finding dotnet version: {e}"
+                        return failwith "Error finding dotnet version, do you have dotnet installed and on the PATH?"
+                    | Ok sdkVersion ->
+                        printfn "Parsed SDK version at root path: %s" sdkVersion.raw
+                        let sdkTfm = tfmForSdkVersion sdkVersion
+                        printfn "Parsed SDK version to tfm: %s" sdkTfm
+                        let fsacTfm, fsacPath = fsacPathForTfm sdkTfm
+                        printfn "Parsed TFM to fsac path: %s" fsacPath
+
+                        let userDotnetArgs = "FSharp.fsac.dotnetArgs" |> Configuration.get [||]
+
+                        let hasUserRollForward =
+                            userDotnetArgs
+                            |> Array.tryFindIndex (fun a -> a = "--roll-forward")
+                            |> Option.map (fun _ -> true)
+                            |> Option.defaultValue false
+
+                        let hasUserFxVersion =
+                            userDotnetArgs
+                            |> Array.tryFindIndex (fun a -> a = "--fx-version")
+                            |> Option.map (fun _ -> true)
+                            |> Option.defaultValue false
+
+                        let shouldApplyImplicitRollForward =
+                            not (hasUserFxVersion || hasUserRollForward)
+                            && sdkTfm <> fsacTfm // if the SDK doesn't match one of our FSAC TFMs, then we're in compat mode
```

Comment:
core of the change is here - if the derived 'default' TFM of the SDK in use is not the TFM of the version of FSAC that we just chose (e.g. net8.0 on a 8.0.100-preview SDK with a net7.0 TFM FSAC binary), then we _must_ apply rollforward flags to force the runtime to choose 8.0 bits. otherwise on a system with 7.0 bits installed we'll use those.